velox/functions/lib/TimeUtils.h

```diff
@@ -123,4 +124,96 @@ struct InitSessionTimezone {
     timeZone_ = getTimeZoneFromConfig(config);
   }
 };
+
+/// Return day-of-year (DOY) of the first `dayOfWeek` in the year.
+///
+/// `dayOfWeek` is a 1-based weekday number starting with Sunday.
+///   (1 = Sunday, 2 = Monday, ..., 7 = Saturday).
+///
+/// If the `dayOfWeek` is Monday, it returns DOY of first Monday in
+/// the year. The returned DOY is a number from 1 to 7.
+FOLLY_ALWAYS_INLINE
+uint32_t getDayOfFirstDayOfWeek(int32_t y, uint32_t dayOfWeek) {
+  auto firstDay =
+      date::year_month_day(date::year(y), date::month(1), date::day(1));
+  auto weekday = date::weekday(firstDay).c_encoding() + 1;
+
+  int32_t delta = dayOfWeek - weekday;
+  if (delta < 0) {
+    delta += 7;
+  }
+
+  return delta + 1;
+}
+
+/// Return the week year represented by Gregorian calendar for the given year,
+/// month and day.
+///
+/// getWeekYear only works with Gregorian calendar due to limitations in the
+/// date library. As a result, dates before the Gregorian calendar
+/// (1582-10-15) yields mismatched results.
+///
+/// The week that includes January 1st and has 'minimalDaysInFirstWeek' or more
+/// days is referred to as week 1. The starting day of the week is decided by
+/// the `firstDayOfWeek`, which is a 1-based weekday number starting with
+/// Sunday.
+///
+/// For ISO 8601, `firstDayOfWeek` is 2 (Monday) and `minimalDaysInFirstWeek`
+/// is 4. For legacy Spark, `firstDayOfWeek` is 1 (Sunday) and
+/// `minimalDaysInFirstWeek` is 1.
+///
+/// The algorithm refers to the getWeekYear algorithm in openjdk:
+/// https://github.com/openjdk/jdk/blob/d9c67443f7d7f03efb2837b63ee2acc6113f737f/src/java.base/share/classes/java/util/GregorianCalendar.java#L2058
```

Comment:
@ccat3z @rui-mo is this different from the vendored date library we have at:

> https://github.com/facebookincubator/velox/blob/main/velox/external/date/iso_week.h

we use it's iso_week functionality in a few places already. If it's the same logic, it would be nice to consolidate the usage.

Reply:
iso_week only support iso 8601 standard week year ([iso_week.h#L1519](https://github.com/facebookincubator/velox/blob/224e4ca5655d3b4fca7384ff9c6a5153879bfe65/velox/external/date/iso_week.h#L1519)). The purpose of this implementation is to support non-ISO standard week year, which is required by Spark.

Reply:
@assignUser @ccat3z @rui-mo I asked for guidance from license experts on this. Give me a day or two until they get back to me.

Reply:
@pedroerp Any updates?